### PR TITLE
[TreeView] Migrate remaining icon props to slots

### PR DESCRIPTION
**Pull request by flaviendelangle**
## Why is this so complex?

I would be very curious to understand the reasoning behind all those icon props.

The table below summarize the available icons and if they can be defined at the Tree View and/or Tree Item level.
Note that there is always one icon rendered, so it's purely a matter of priority between them:

|                | Description                                                                                                                                     | Tree View | Tree Item |
|----------------|-------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------|
| `icon`         | Always rendered                                                                                                                                 |           | ✅         |
| `expandIcon`   | Rendered if the element has children and it is collapsed (except if `icon` is defined)                                                          | ✅         | ✅         |
| `collapseIcon` | Rendered if the element has children and it is expanded (except if `icon` is defined)                                                           | ✅         | ✅         |
| `parentIcon`   | Rendered if the element has children<br />(so only if `icon`, `expandIcon` and `collapseIcon` are null both on the Tree View and the Tree Item) | ✅         |           |
| `endIcon`      | Rendered if the element has no children (except if `icon` is defined)                                                                           | ✅         | ✅         |

### `parentIcon`

This highlight a small behavior change caused by #11664
Now, if you just pass a `parentIcon` to your Tree View, it is not used since you always have a `collapseIcon` and `expandIcon` defined.

IMHO this behavior is waaaaaay to complex for the actual benefit.
I would at the very least be in favor of dropping `parentIcon`, AFAIK the two following components have always been equivalent:

```tsx
<TreeView defaultEndIcon={<SomeIcon />} />
<TreeView defaultExpandIcon={<SomeIcon />} defaultCollapseIcon={<SomeIcon />} />
```

We would end up with 4 icons (which is still a lot) with the following behavior:

```
If `icon` is defined, then use `icon`
Else, if the node has no children, then use `endIcon`
Else, if the node is expanded, then use `collapseIcon`,
Else, use `expandIcon`
```

All icons, would be defined both on the Tree View and on the Tree Item, except for `icon` which is here to customize some specific nodes (and should be documented by the way, the only time we have a use case for it, we set the custom icon **inside** the content, in the Gmail clone, so we don't use this slot)

### CODE REVIEW

File: docs/data/migration/migration-tree-view-v6/migration-tree-view-v6.md [hunk 1]
```diff
@@ -178,6 +178,81 @@ you need to use the new `collapseIcon` slot on this component:
...
+-     endIcon={<MyCustomEndIcon />}
++     slots={{ endIcon: MyCustomEndIcon }}
+    />
+  </SimpleTreeView>
+```
+
+#### Define `icon`
+
+The icon rendered next to item
+is now defined as a slot both on the Tree View component.
```
Comment (michelengelen) [inline docs/data/migration/migration-tree-view-v6/migration-tree-view-v6.md:236]:
```suggestion
is now defined as a slot both on the Tree View and the Tree Item component.
```

**Review by michelengelen — APPROVED** [review]

File: docs/data/migration/migration-tree-view-v6/migration-tree-view-v6.md [hunk 2]
```diff
@@ -178,6 +178,81 @@ you need to use the new `collapseIcon` slot on this component:
...
+
+If you were passing an icon to your Tree Item component,
+you need to use the new `endIcon` slot on this component:
```
Comment (michelengelen) [inline docs/data/migration/migration-tree-view-v6/migration-tree-view-v6.md:239]:
```suggestion
you need to use the new `icon` slot on this component:
```

**Review by michelengelen — APPROVED** [review]
found one more i missed before 🙇🏼

File: packages/x-tree-view/src/internals/plugins/useTreeViewIcons/useTreeViewIcons.types.ts [hunk 3]
```diff
@@ -28,19 +17,29 @@ interface UseTreeViewIconsSlots {
    * The default icon used to expand the node.
    */
   expandIcon?: React.ElementType;
+  /**
+   * The default icon displayed next to an end node.
+   * This is applied to all tree nodes and can be overridden by the TreeItem `icon` prop.
```
Comment (LukasTy) [inline packages/x-tree-view/src/internals/plugins/useTreeViewIcons/useTreeViewIcons.types.ts:22]:
```suggestion
   * This is applied to all tree nodes and can be overridden by the TreeItem `icon` slot prop.
```

File: packages/x-tree-view/src/internals/plugins/useTreeViewIcons/useTreeViewIcons.types.ts [hunk 4]
```diff
@@ -28,19 +17,29 @@ interface UseTreeViewIconsSlots {
...
+   */
+  endIcon?: React.ElementType;
+  /**
+   * The default icon displayed next to a parent node.
+   * This is applied to all parent nodes and can be overridden by the TreeItem `icon` prop.
```
Comment (LukasTy) [inline packages/x-tree-view/src/internals/plugins/useTreeViewIcons/useTreeViewIcons.types.ts:27]:
```suggestion
   * This is applied to all parent nodes and can be overridden by the TreeItem `endIcon` slot prop.
```

**Review by LukasTy — APPROVED** [review]

File: docs/translations/api-docs/tree-view/tree-item/tree-item.json [hunk 5]
```diff
@@ -9,8 +9,6 @@
     },
     "ContentProps": { "description": "Props applied to ContentComponent." },
     "disabled": { "description": "If <code>true</code>, the node is disabled." },
-    "endIcon": { "description": "The icon displayed next to an end node." },
```
[the remark this answers is not in the document]
Comment (flaviendelangle) [reply docs/translations/api-docs/tree-view/tree-item/tree-item.json:12]: Indeed, I'll have a check on the infra to do a follow up :+1:

**Review by noraleonte — APPROVED** [review]
Looks great! 🎉  I'm also in favor of dropping the `parentIcon`, I don't really see any value in it 🤔

**Comment by flaviendelangle** [comment]
@noraleonte @LukasTy I removed the `parentIcon` slot, I let you have a look
I did not change the behavior in `TreeItem` to avoid breaking changes, but it's obvious that once we create a clean headless approach for the Tree Item, we should no longer pass those 3 props `icon` `displayIcon` and `expansionIcon` to the content.

File: docs/translations/api-docs/tree-view/tree-item/tree-item.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (LukasTy) [reply docs/translations/api-docs/tree-view/tree-item/tree-item.json:12]: Nice. 💯

**Pull request merged**